### DISCUSSION TITLE
NormalizeFeatures with negative values?

### DISCUSSION THREAD

**Discussion by Shualdon** [pyg-team/pytorch_geometric#4668] [Q&A]
Hi,

I'm trying to normalize the features in my data set and some of the features I use are negative. I noticed that the NormalizeFeatures transform doesn't work if there are any negative values in the input tensor. 
for example, if I run my original tensor (with one negative value) I get the same tensor in return:

```python
x = torch.tensor([[3.5411, 0.1826, 0.4144, 0.9658, 0.4737, 55.3373, 0.0000]], dtype=torch.float)

data = Data(x=x)
print(data.x)
data = NormalizeFeatures()(data)
print(data.x)
 
>> tensor([[  3.5411,   0.1826,   0.4144,   0.9658,   0.4737, -55.3373,   0.0000]])
>> tensor([[  3.5411,   0.1826,   0.4144,   0.9658,   0.4737, -55.3373,   0.0000]])
```

But if I change the sign for the negative value I get a normalized tensor:

```python
x = torch.tensor([[3.5411, 0.1826, 0.4144, 0.9658, 0.4737, 55.3373, 0.0000]], dtype=torch.float)

data = Data(x=x)
print(data.x)
data = NormalizeFeatures()(data)
print(data.x)

>> tensor([[ 3.5411,  0.1826,  0.4144,  0.9658,  0.4737, 55.3373,  0.0000]])
>> tensor([[0.0581, 0.0030, 0.0068, 0.0159, 0.0078, 0.9084, 0.0000]])
```

Is there another transform that can normalize the features to a [-1,1] range instead of [0,1]?

**Comment by rusty1s** [comment]
It looks like this is already fixed in PyG 2.0.4. Can you upgrade to confirm?

If you want to normalize features to range `[-1, 1]`, you can do so easily by simply writing your own transform :)

**Reply by Shualdon** [reply]
Upgraded and now it works, but I see it just zeroes out the negative values. I guess I'll have to write my own transform :)

**Reply by rusty1s** [reply]
It does not zero out the negatives, it will map it to range `[0, 1]`such that `x.min()` gets mapped to `0` and `x.max()` gets mapped to `1`.

**Reply by Shualdon** [reply]
Yeah, I figure it out after I sent the message. Thanks :)
